aspnetcore/fundamentals/dependency-injection.md

```diff
@@ -692,3 +544,538 @@ DI is an *alternative* to static/global object access patterns. You may not be a
+* [Grace](https://www.nuget.org/packages/Grace.DependencyInjection.Extensions)
+* [LightInject](https://github.com/seesharper/LightInject.Microsoft.DependencyInjection)
+* [Lamar](https://jasperfx.github.io/lamar/)
+* [Stashbox](https://github.com/z4kn4fein/stashbox-extensions-dependencyinjection)
+* [Unity](https://www.nuget.org/packages/Unity.Microsoft.DependencyInjection)
+
+### Thread safety
+
+Create thread-safe singleton services. If a singleton service has a dependency on a transient service, the transient service may also require thread safety depending how it's used by the singleton.
+
+The factory method of single service, such as the second argument to [AddSingleton\<TService>(IServiceCollection, Func\<IServiceProvider,TService>)](xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton*), doesn't need to be thread-safe. Like a type (`static`) constructor, it's guaranteed to be called once by a single thread.
+
+## Recommendations
+
+* `async/await` and `Task` based service resolution is not supported. C# does not support asynchronous constructors; therefore, the recommended pattern is to use asynchronous methods after synchronously resolving the service.
+
+* Avoid storing data and configuration directly in the service container. For example, a user's shopping cart shouldn't typically be added to the service container. Configuration should use the [options pattern](xref:fundamentals/configuration/options). Similarly, avoid "data holder" objects that only exist to allow access to some other object. It's better to request the actual item via DI.
+
+* Avoid static access to services (for example, statically-typing [IApplicationBuilder.ApplicationServices](xref:Microsoft.AspNetCore.Builder.IApplicationBuilder.ApplicationServices) for use elsewhere).
+
+* Avoid using the *service locator pattern*. For example, don't invoke <xref:System.IServiceProvider.GetService*> to obtain a service instance when you can use DI instead:
+
+  **Incorrect:**
+
+  ```csharp
+  public class MyClass()
+  {
+      public void MyMethod()
+      {
+          var optionsMonitor = 
+              _services.GetService<IOptionsMonitor<MyOptions>>();
+          var option = optionsMonitor.CurrentValue.Option;
+
+          ...
+      }
+  }
+  ```
+
+  **Correct**:
+
+  ```csharp
+  public class MyClass
+  {
+      private readonly IOptionsMonitor<MyOptions> _optionsMonitor;
+
+      public MyClass(IOptionsMonitor<MyOptions> optionsMonitor)
+      {
+          _optionsMonitor = optionsMonitor;
+      }
+
+      public void MyMethod()
+      {
+          var option = _optionsMonitor.CurrentValue.Option;
+
+          ...
+      }
+  }
+  ```
+
+* Another service locator variation to avoid is injecting a factory that resolves dependencies at runtime. Both of these practices mix [Inversion of Control](/dotnet/standard/modern-web-apps-azure-architecture/architectural-principles#dependency-inversion) strategies.
```

Comment:
Do you specifically mean a factory that **uses a service locator** to resolve dependencies? Otherwise, this could be misconstrued with the factory form given in the example above and may seem contradictory to the guidance in the Disposal section:

```services.AddSingleton<IService3>(sp => new Service3());```

Reply:
I don't think it's a significant problem following what was just shown with "Another service locator variation ...", but this sentence could say ...

> Another service locator variation to avoid is injecting a factory that resolves dependencies at runtime using \<xref:System.IServiceProvider.GetService*>.

Reply:
This does add clarity IMO. 👍